### PR TITLE
Upgrade Flow: Connecting Jetpack users

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
If a Jetpack user is not connected to WP.com, clicking on the "Upgrade" button will lead them nowhere.
This PR introduces the changes to let such users register/authenticate on WP.com (if needed), and link their account to the Jetpack site.

#### Jetpack product discussion
p9dueE-1M4-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

**Due to changes in https://github.com/Automattic/jetpack/pull/18548, this PR (as well as the `master` branch itself) outputs error notices.
You'll need to disable `WP_DEBUG` for redirects to work properly.**

1. Connect a Jetpack site with a Free plan, create an additional WP user.
2. Log in as that user in another browser or a private mode. If you're logged in WP.com in that browser, please logout.
3. Click an "Upgrade" button in Jetpack Dashboard or Jetpack Settings. You should be redirected to the Calypso connection flow.
4. Authenticate into WP.com. You should be redirected to the WP.com checkout page with that product in your cart.
5. Go back to Jetpack Dashboard, confirm that you are connected as a secondary user. Unlink that user from WP.com.
6. Click on any other "Upgrade" button. You should be redirected to the Calypso connection flow being logged into WP.com.
7. After clicking "Approve", you should get redirected to WP.com checkout with that product in the cart (previous "upgrade" product will probably still be in the cart too).
8. Go back to Jetpack Dashboard, confirm that you're connected to WP.com as a secondary user.

#### Proposed changelog entry for your changes:
* Improved product upgrade flow for non-connected Jetpack users.

### CODE REVIEW

File: class.jetpack.php [hunk 1]
```diff
@@ -4178,6 +4178,32 @@ function admin_page_load() {
 
 		if ( isset( $_GET['action'] ) ) {
 			switch ( $_GET['action'] ) {
+				case 'authorize_redirect':
+					self::log( 'authorize_redirect' );
+
+					add_filter( 'allowed_redirect_hosts', function( $domains ) {
```
Comment (ebinnion) [inline class.jetpack.php:4184]: Are anonymous functions allowed in our supported PHP versions?

**Review by leogermani — CHANGES_REQUESTED** [review]
Very well done. Works as expected in various situations

* not connected secondary user
* totally user-less connection

I also tested that the user is properly connected even If the checkout is not concluded,

I just found one small problem:

In the calypso screen that asks for the user authorization, if you give up authorizing and click "Back to __sitename__", you end up in an endless loop, because our new redirect sends us back to Calypso again.

**Comment by sergeymitr** [comment]
> In the calypso screen that asks for the user authorization, if you give up authorizing and click "Back to sitename", you end up in an endless loop, because our new redirect sends us back to Calypso again.

That's a good catch, thanks @leogermani.
I've pushed the fix, please re-review.

**Review by leogermani — APPROVED** [review]
tested again and everything is looking fine!

**Pull request merged**